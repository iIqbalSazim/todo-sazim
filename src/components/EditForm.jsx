import {
  Button,
  Center,
  Modal,
  Select,
  SimpleGrid,
  Textarea,
  Title,
} from "@mantine/core";
import { DateInput } from "@mantine/dates";
import { selectPriorityData } from "../constants/formData";
import { formatDueDate } from "../helper/helper";
import { BUTTON } from "../constants/colors";
import { useState } from "react";

const EditForm = ({ task, editTask, isOpen, closeModal }) => {
  const [editedTask, setEditedTask] = useState({ ...task });

  const handleSubmit = (e) => {
    e.preventDefault();

    editTask(editedTask);
    closeModal();
  };

  return (
    <Modal opened={isOpen} onClose={() => closeModal()} centered>
      <form onSubmit={handleSubmit}>
        <SimpleGrid cols={1} p={"lg"}>
          <Title ta={"center"}>Edit Todo</Title>
          <Textarea
            defaultValue={editedTask.description}
            onChange={(e) =>
              setEditedTask({ ...editedTask, description: e.target.value })
            }
          />
          <Select
            data={selectPriorityData}
            defaultValue={editedTask.priority}
            onSelect={(e) =>
              setEditedTask({
                ...editedTask,
                priority: e.target.value.toLowerCase(),
              })
            }
          />
          <DateInput
            minDate={new Date()}
            defaultValue={new Date(editedTask.dueDate)}
            onChange={(input) =>
              setEditedTask({ ...editedTask, dueDate: formatDueDate(input) })
            }
            label="Date input"
            placeholder="Date input"
          />
          <Center>
            <Button color={BUTTON} type="submit">
              Update
            </Button>
          </Center>
        </SimpleGrid>
      </form>
    </Modal>
  );
};

export default EditForm;
