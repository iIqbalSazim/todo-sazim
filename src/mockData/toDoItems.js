export const mockToDoItems = [
  {
    id: 1,
    description: "Test task 1",
    priority: "medium",
    created_at: "2023-11-30T09:45:36.302Z",
    due_date: "2023-12-22T18:00:00.000Z",
    is_completed: false,
    is_deleted: false,
  },
  {
    id: 2,
    description: "Test task 2",
    priority: "high",
    created_at: "2023-11-27T11:05:37.765Z",
    due_date: "2023-12-30T18:00:00.000Z",
    is_completed: false,
    is_deleted: false,
  },
  {
    id: 3,
    description: "Test task 3",
    priority: "low",
    created_at: "2023-11-30T12:33:57.522Z",
    due_date: "2023-11-29T18:00:00.000Z",
    is_completed: false,
    is_deleted: false,
  },
];

export const mockToDo = {
  id: 1,
  description: "Task description",
  priority: "high",
  due_date: "2023-12-31T00:00:00.000Z",
};

export const mockTrashTodos = [
  {
    id: 1,
    description: "Test task 1",
  },
  {
    id: 2,
    description: "Test task 2",
  },
];
