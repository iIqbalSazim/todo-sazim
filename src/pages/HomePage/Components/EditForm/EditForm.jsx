import { useState } from "react";

import {
  Button,
  Center,
  Modal,
  Select,
  SimpleGrid,
  Textarea,
  Title,
} from "@mantine/core";
import { DateInput } from "@mantine/dates";

import { PRIORITY_OPTIONS, COLORS } from "../../HomePageConstants";
import { updateTask } from "../../Api/Methods";

const EditForm = ({ task, isOpen, closeModal, editTask }) => {
  const [editedTask, setEditedTask] = useState({ ...task });

  const handleSubmit = async (e) => {
    e.preventDefault();

    await editTask(editedTask);

    closeModal();
    alert("Task successfully updated");
  };

  return (
    <Modal opened={isOpen} onClose={() => closeModal()} centered>
      <form onSubmit={handleSubmit}>
        <SimpleGrid cols={1} p={"lg"}>
          <Title ta={"center"}>Edit Todo</Title>
          <Textarea
            defaultValue={editedTask.description}
            onChange={(e) =>
              setEditedTask({ ...editedTask, description: e.target.value })
            }
          />
          <Select
            data={PRIORITY_OPTIONS}
            defaultValue={editedTask.priority}
            onSelect={(e) =>
              setEditedTask({
                ...editedTask,
                priority: e.target.value.toLowerCase(),
              })
            }
          />
          <DateInput
            minDate={new Date()}
            defaultValue={new Date(editedTask.due_date)}
            onChange={(input) =>
              setEditedTask({ ...editedTask, due_date: input })
            }
            label="Due Date"
          />
          <Center>
            <Button color={COLORS.BUTTON} type="submit">
              Update
            </Button>
          </Center>
        </SimpleGrid>
      </form>
    </Modal>
  );
};

export default EditForm;
