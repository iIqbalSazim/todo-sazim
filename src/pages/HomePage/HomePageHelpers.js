export function findIndexWithId(data, id) {
  return data.findIndex((element) => element.id === id);
}

export function generateCurrentTimeAndDate(date) {
  let currentTime =
    date.toLocaleTimeString().slice(0, 4) + date.toLocaleTimeString().slice(7);
  let currentDate = `${currentTime} | ${date.toDateString()}`;
  return currentDate;
}

export function formatDueDate(date) {
  return date.toDateString().slice(0, 15);
}
