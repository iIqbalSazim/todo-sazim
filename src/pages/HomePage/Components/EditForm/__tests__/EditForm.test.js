import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { mockToDo } from "../../../../../mockData/toDoItems";

import EditForm from "../EditForm";
import { replaceAllRandomGeneratedIds } from "../../../HomePageHelpers";

describe("EditForm", () => {
  const mockEditTask = jest.fn();

  test("renders correctly with the correct task data", async () => {
    const { container } = render(
      <EditForm
        task={mockToDo}
        isOpen={true}
        closeModal={() => {}}
        editTask={mockEditTask}
      />
    );

    const titleElement = screen.getByText("Create Todo");
    const taskInputElement = screen.getByPlaceholderText("Write your task");
    const setPriorityElement = screen.getByRole("searchbox");
    const dueDateElement = screen.getByLabelText("Due Date");
    const updateButtonElement = screen.getByRole("button", {
      name: "Update",
    });

    expect(titleElement).toBeInTheDocument();
    expect(taskInputElement).toBeInTheDocument();
    expect(setPriorityElement).toBeInTheDocument();
    expect(dueDateElement).toBeInTheDocument();
    expect(updateButtonElement).toBeInTheDocument();

    expect(taskInputElement).toHaveValue("Task description");
    expect(setPriorityElement).toHaveValue("High");
    expect(dueDateElement).toHaveValue("December 31, 2023");

    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  });

  test("calls editTask with the correct task data", async () => {
    const { container } = render(
      <EditForm
        task={mockToDo}
        isOpen={true}
        closeModal={() => {}}
        editTask={mockEditTask}
      />
    );

    const taskInputElement = screen.getByPlaceholderText("Write your task");
    const dueDateElement = screen.getByLabelText("Due Date");
    const updateButtonElement = screen.getByRole("button", {
      name: "Update",
    });

    await userEvent.clear(taskInputElement);
    await userEvent.type(taskInputElement, "Updated Task Description");

    await userEvent.clear(dueDateElement);
    await userEvent.type(dueDateElement, "December 24, 2023");

    await userEvent.click(updateButtonElement);

    await waitFor(() => {
      expect(mockEditTask).toHaveBeenCalledTimes(1);
      expect(mockEditTask).toHaveBeenCalledWith({
        id: 1,
        description: "Updated Task Description",
        priority: "high",
        due_date: new Date("2023-12-23T18:00:00.000Z"),
      });
    });

    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  }, 20000);
});
