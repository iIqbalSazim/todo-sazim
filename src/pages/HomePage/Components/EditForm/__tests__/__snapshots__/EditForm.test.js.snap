// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm calls editTask with the correct task data 1`] = `""`;

exports[`EditForm renders correctly with the correct task data 1`] = `""`;
